### ISSUE TITLE
Completion of strings and dict keys

### ISSUE THREAD

**Issue by jnothman** [davidhalter/jedi#771]
I [implemented](https://github.com/ipython/ipython/pull/5304) completion of dict keys (and Pandas columns) for IPython and continue to find it extremely useful. IPython since then [added](https://github.com/ipython/ipython/pull/9289) support for an `_ipy_key_completions_` method on custom classes. You are welcome to use my code for key completion to support similar functionality here.

But, in the same way that Jedi adopts type inference from the context, strings found in one's context are often reused. Completion of string literals already used in a session is a novel idea worth trying.

**Comment by davidhalter** [comment]
Well, I'm open for a discussion. I'm happy to discuss well thought out proposals and pull requests :)

However your code doesn't help with that at all. Jedi has a whole other levels of abstractions that you need to fight through. Thinking about Jedi means thinking about both evaluated code (like ipython) and parsed code.

**Comment by jnothman** [comment]
When we're talking about dict keys, non-evaluated code is irrelevant. You need to actually inspect the particular object the user has their cursor following: it's rarely a function of type. I am suggesting this should be a feature outside of Jedi, just as path completion is, unless you think it clearly falls within Jedi's scope.

Regarding my code: the challenges in doing string completion involve things like string quoting, escaping, encoding, etc. This is not, to my knowledge, covered by Jedi.

**Comment by davidhalter** [comment]
> When we're talking about dict keys, non-evaluated code is irrelevant.

Why?

> This is not, to my knowledge, covered by Jedi.

This is all covered by the syntax tree and the tokenizer.

**Comment by jnothman** [comment]
As far as I'm concerned you can't work out what keys an object has without evaluating that object's reference.

Perhaps I've not understood Jedi well enough yet. If you feel this is a feature for Jedi, please, direct it there.

**Comment by davidhalter** [comment]
I just fixed #951. This means that dict completions work very well now. 

Now what's potentially left to do is to add something like  `_ipy_key_completions_`. However, by searching with multiple tools I only found this one reference where it wasn't even merged: https://github.com/pybind/pybind11/pull/1405. So it looks like this is not really a feature that's used.

I'm closing, because I don't feel anybody would be using this. I'm happy to reopen and implement it if pandas core developers are willing to add something like `_ipy_key_completions_` to their code. It's basically the only tool where something like this would be worth it.

Below you see my WIP patch (some test cases are already passing):

```
diff --git a/jedi/inference/compiled/access.py b/jedi/inference/compiled/access.py
index d76fad8..dabe8aa 100644
--- a/jedi/inference/compiled/access.py
+++ b/jedi/inference/compiled/access.py
@@ -240,11 +240,27 @@ class DirectObjectAccess(object):
         return tuple(self._create_access_path(cls) for cls in self._obj.__mro__[1:])
 
     def py__getitem__all_values(self):
+        if self.is_instance():
+            try:
+                func = getattr(self._obj, '__jedi_get_dct__')
+            except AttributeError:
+                pass
+            else:
+                return [self._create_access_path(v) for v in func().values()]
+
         if isinstance(self._obj, dict):
             return [self._create_access_path(v) for v in self._obj.values()]
         return self.py__iter__list()
 
     def py__simple_getitem__(self, index):
+        if self.is_instance():
+            try:
+                func = getattr(self._obj, '__jedi_get_dct__')
+            except AttributeError:
+                pass
+            else:
+                return self._create_access_path(func()[index])
+
         if type(self._obj) not in ALLOWED_GETITEM_TYPES:
             # Get rid of side effects, we won't call custom `__getitem__`s.
             return None
diff --git a/test/test_api/test_interpreter.py b/test/test_api/test_interpreter.py
index e91523e..4059102 100644
--- a/test/test_api/test_interpreter.py
+++ b/test/test_api/test_interpreter.py
@@ -615,10 +615,31 @@ def test_dict_completion(code, column, expected):
         ('dct["asdf"]', ['float']),
         ('dct[r"asdf"]', ['float']),
         ('dct["a"]', ['float', 'int']),
+
+        ('Dct1["lala"]', ['Dct1']),  # Basically generics
+        ('Dct2["lala"]', []),
+
+        ('dct1["lala"]', ['float']),
+        ('dct1["la"]', ['float', 'str']),
+        ('dct2["lala"]', ['float']),
+        ('dct2["la"]', ['float', 'str']),
     ]
 )
 def test_dict_getitem(code, types):
     dct = {1: 2, "asdf": 1.0}
 
+    class Dct1():
+        def __jedi_get_dct__(self):
+            return {'lala': 3.0, 'fufu': ""}
+
+    class Dct2():
+        def __jedi_get_dct__(self):
+            return {'lala': 3.0, 'fufu': ""}
+
+    Dct2.__name__ = 'Bra'
+
+    dct1 = Dct1()
+    dct2 = Dct2()
+
     comps = jedi.Interpreter(code, [locals()]).infer()
     assert [c.name for c in comps] == types
```

**Comment by yangle** [comment]
> Now what's potentially left to do is to add something like `_ipy_key_completions_`. However, by searching with multiple tools I only found this one reference where it wasn't even merged

That's because this method has been renamed to `_ipython_key_completions_` since forever ago. :)
